### PR TITLE
[v3] remove _handleSet from _TemporalState

### PR DESCRIPTION
**Pull request by pstrassmann**
This PR removes the `_handleSet` internal property of `_TemporalState` in favor of defining `handleSet` within `temporal` (index.ts). 

This is beneficial from a complexity-reducing standpoint: `_handleSet` no longer needs to be a member of the temporal state, and its functionality is easily discoverable and readable within `temporal`.

  This change also allows us to remove the type assertion required for accessing private internal properties of `_TemporalState` when using `getState()`, which returns type `TemporalState`.

### CODE REVIEW

**Comment by charkour** [comment]
Ah, I see the issue. This is a really annoying part of zustand and I don't know the best practice. I'll make a discussion on the zustand repo to see what the problem is. 

I _think_ the tests are failing due to a bug. We override the `set` callback in the `wrapTemporal` test but not the `store.setState`. If you add `store.setState = set;` right before returning the `config`, that should "fix" the tests. I think it's expected the middleware modify both, but I need to check with the zustand maintainers. 

image

**Comment by charkour** [comment]
Made the discussion here: https://github.com/pmndrs/zustand/discussions/2305

**Comment by pstrassmann** [comment]
Ah okay thank you for this! To make sure I understand, is the potential bug of directly modifying setState due to the fact that if multiple nested middlewares do this, setState will just be identical to the last middleware-defined set function instead of applying all nested middleware's set functionality? Or  is it that if not all middleware's modify setState, set and setState would differ? I may also still be a little confused...

**Comment by charkour** [comment]
> Or is it that if not all middleware's modify setState, set and setState would differ?

Yup, exactly! If the user of the middleware expects that set and setState behave the same, then it's an issue. 

Looking at this more, I think we'll want to move forward with this change, but it'll have to be for v3 because it's a breaking change. It depends on setState rather than set.

**Comment by pstrassmann** [comment]
Ok sounds good! I updated the tests with your suggested fixes, which indeed stopped the tests from failing. If there's anything else you'd like me to do with this PR, just let me know.

File: tests/__tests__/options.test.ts
Comment (charkour) [inline tests/__tests__/options.test.ts:704]: Hey @pstrassmann, sorry for the delay. Life has been busy and I haven't made time for working on v3 of zundo; however, I'm here now after some good [discussions in the zustand repo regarding authoring middleware](https://github.com/pmndrs/zustand/discussions/2305). 

Before merging this in, could you add another test case where `store.setState = set` isn't added and we expect `.toHaveBeenCalled(0)`? To put it in English, we should have a test case for zundo, there `wrapTemporal` fails to be called because we haven't modified the store setter.

File: src/index.ts
Comment (charkour) [inline src/index.ts:74]: Why is this line needed here? Shouldn't we only need to be passing `pastStates` and `futureStates` here?

Comment (pstrassmann) [reply src/index.ts:74]: I (mistakenly, I think) was under the impression that `setState` did not do the kind of auto-merging of state that `set` does, which is why I spread the other values here. I will remove this line.

File: tests/__tests__/options.test.ts
Comment (pstrassmann) [inline tests/__tests__/options.test.ts:663]: Please feel free to suggest a better test description, or if you'd prefer the commented out lines to be removed

Comment (charkour) [reply tests/__tests__/options.test.ts:663]: Awesome, thank you!

File: src/index.ts
Comment (charkour) [reply src/index.ts]: Gotcha, thanks. I _think_ they both behave the same.

**Review by charkour — APPROVED** [review]
Thanks for the great work on this!

**Pull request merged**
